File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.js
function solution(a, b, c, d) {
    let answer = 0;
    const obj = {};
    [a, b, c, d].forEach(num => obj[num] = (obj[num] || 0) + 1);
    
    const keys = Object.keys(obj);
    const len = keys.length;
    
    if(len === 1) {
        const p = keys[0];
        answer = 1111 * p;
    } else if(len === 2) {
        const [p, q] = keys;
        console.log(p, q)
        if(obj[p] > obj[q]) {
            answer = (10 * +p + +q) ** 2;    
        } else if(obj[p] < obj[q]) {
            answer = (10 * +q + +p) ** 2;
        } else {
            answer = (+p + +q) * Math.abs(+p - +q);
        }
    } else if (len === 3) {
        const p = keys[0];
        const q = keys[1];
        const r = keys[2];
        const num1 = obj[keys[0]];
        const num2 = obj[keys[1]];
        if(num1 > num2) {
            answer = q * r;
        } else if (num1 < num2){
            answer = p * r;
        } else {
            answer = p * q;
        }
    } else {
        answer = Math.min(...keys);
    }
    
    
    return answer;
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.py
def solution(a, b, c, d):
    dic = {}
    arr = [a,b,c,d]
    for i in arr:
        dic[i] = dic.get(i,0) + 1
    l = len(dic)
    
    result = sorted(dic.items(), key=lambda x: x[1], reverse=True)
    values = sorted(list(set([v for k,v in result])), reverse=True)
    keys = [k for k, v in result]
    print(result, l, values)
    if l == 1: return 1111 * keys[0]
    if l == 4: return min(keys)
    if l == 2: 
        if len(values) == 1: return (keys[0] + keys[1]) * abs(keys[0] - keys[1])
        else: return (keys[0] * 10 + keys[1])**2
    if l == 3: return keys[1] * keys[2]
        
        
    